refactor!: Replace premake5 with CMake.

tupoy-ya opened the pull request:
# Reason
I want to make YimMenu MinGW compatible and it's much easier to do with CMake.

Related:
+ https://github.com/YimMenu/YimMenu/pull/342#issuecomment-1268391294

# Additional context
- [x] Compiles
- [x] Runs
- [x] CI
- [x] Test with v142 tool set
- [x] Test with v143 tool set
- [x] Update `README.md`
- [x] Make everything in vendor directory appear in vs
- [x] Update script.bat
- [ ] Maybe remove some `CMakePredefinedTargets`
- [ ] Maybe use `YimMenu/bin` as output directory
- [ ] Maybe remove Debug build
- [ ] Don't increase amount of people failing to build YimMenu by 90%
- [x] Warnings as errors

tupoy-ya committed:
refactor!: Replace premake5 with CMake.

tupoy-ya committed:
fix: Submodules.

tupoy-ya committed:
fix: CI. I hope.
chore(logger.cpp): Remove <format> include.
refactor(CMakeLists.txt): Tabs to spaces.

tupoy-ya committed:
fix(CI): Maybe i should remove bash?

tupoy-ya committed:
fix(CI): I'm blind.
working-directory: build/Release/

GeopJr reviewed (changes requested):
Could you also get rid of gitsubmodules and instead use [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html)?

tupoy-ya committed:
refactor(CMakeLists.txt)!: Replace git submodules with FetchContent.
feat(CMakeLists.txt): Remove debug build.

tupoy-ya committed:
chore(README.md): Update README.md to have manual updates and CMake.

tupoy-ya committed:
chore(CMakeLists.txt): Warnings as errors.

tupoy-ya committed:
fix(script.bat): CMake support, --build option.

tupoy-ya committed:
fix(script.bat): Remove `--build` option i hate DOS and windows.

tupoy-ya committed:
chore(CMakeLists.txt): Clean up.

tupoy-ya committed:
refactor!: Clean up CMakeLists.txt and move fetch content into includes.
chore(README.md): Fix typo.

tupoy-ya committed:
fix: C++ standard and warning 8028.

tupoy-ya committed:
chore(CMakeLists.txt): Remove detected `MinSizeRel`.
fix(g3log): Ignoring options.

GeopJr commented on CMakeLists.txt:
```diff
@@ -0,0 +1,60 @@
+include(scripts/json.cmake)
+
+message("\nFetching custom modules")
+include(scripts/ImGui.cmake)
```
```suggestion
include(scripts/imgui.cmake)
```

case-sensitive (at least on unix)

tupoy-ya committed:
fix(CMake scripts): All scripts to lowercase.

Yimura commented on CMakeLists.txt:
```diff
@@ -0,0 +1,60 @@
+target_link_libraries(BigBaseV2 PRIVATE pugixml minhook g3log imgui d3dcompiler)
+
+# Disable warnings
+set_target_properties(BigBaseV2 PROPERTIES COMPILE_FLAGS "/wd4996") # C4996: deprecated in C++17
```
Where does this warning occur? Let's try to fix it instead of silencing it, we're on C++20 anyways.

tupoy-ya replied:
This warning doesn't occur anymore but did when i added this.

Yimura commented on BigBaseV2/src/main.cpp:
```diff
@@ -31,10 +31,10 @@ BOOL APIENTRY DllMain(HMODULE hmod, DWORD reason, PVOID)
 
 		g_hmodule = hmod;
-		g_is_steam = GetModuleHandle(L"steam_api64.dll") != NULL;
+		g_is_steam = GetModuleHandle("steam_api64.dll") != NULL;
```
We can remove this as all there only exists one version of GTA currently (since b2699).

Yimura reviewed (changes requested):
I'm fine with keeping the `build/` folder instead of `bin/`.

Regarding build types `Debug/Release/MinSizeRel/RelWithDebInfo`, keep them all.
It should be up to the developer to choose the correct one.

Since we're doing this big change, why don't we remove the top-level `BigBaseV2/` folder and just straight up go with `src/`.
Continuing this, YimMenu is barely anything like BBv2, how about we switch the name in the CMake project to YimMenu instead?
Possibly also changing the README title as it is misleading and instead reference BBv2 as our "roots" somewhere further down in the README.

Possibly also remove the `.bat` scripts from the repository, I'll probably be using VS Code instead of Visual Studio from now on and most of the problems that people had were from the submodules which will be removed/handled by CMake instead.

tupoy-ya committed:
chore(CMakeLists.txt): Allow all build types.
chore(CMakeLists.txt): Remove "Disable warnings" (Needs test in CI).
refactor(CMakeLists.txt): Change "BigBaseV2" to "YimMenu".
chore: Remove g_is_steam check.

tupoy-ya committed:
refactor!: Replace `BigBaseV2/src/` with `src/`.
fix(CI): Fix CI trying to access `BigBaseV2.dll` instead of `YimMenu.dll`.

tupoy-ya committed:
chore: Merge branch 'master' of https://github.com/YimMenu/YimMenu into feat-cmake-build.
That was easier then i expected.

tupoy-ya committed:
refactor(README.md): Replace "original"
[BigBaseV2](https://github.com/Pocakking/BigBaseV2) mention with [UC](https://www.unknowncheats.me/forum/grand-theft-auto-v/476972-yimmenu-1-63-a.html) one.
I hope markdown works.

tupoy-ya committed:
chore: Merge branch 'master' of https://github.com/YimMenu/YimMenu into feat-cmake-build.

tupoy-ya committed:
feat(.gitignore): Add `.vscode/` and `.cache`.
chore(Batch scripts): Remove .bat scripts.

tupoy-ya committed:
fix(README.md): Remove script mention.

Yimura reviewed (approved):
Approving for now, if people still have something to say about what they'd like to be done differently have 6h left before I merge.
